Change: Replace fixed length _grf_text array with vector.

PeterN opened the pull request:
## Motivation / Problem

On x86-64, `_grf_text` is a 20MB array. This was due to increasing the number of permitted IDs without realising the ID is used as an array index.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

IDs are allocated sequentially by OpenTTD, not by NewGRFs, so it is possible to replace the array with a vector and allocate as necessary. The limit is still enforced, though is unlikely to be reached.

Additionally reshuffle GRFTextEntry for better alignment. This only seems to help 32bit x86, but shouldn't hurt other architectures.


<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

An assert has been added for out-of-bounds access. I have assumed the NewGRFs cannot interfere and pass IDs directly.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Change: Replace fixed length _grf_text array with vector.

Additionally reshuffle GRFTextEntry for better alignment.

This removes a mostly-unused static 20MB allocation.

glx22 commented on src/newgrf_text.cpp:
```diff
@@ -561,27 +560,22 @@ StringID AddGRFString(uint32_t grfid, uint16_t stringid, byte langid_to_add, boo
+
+		/* We didn't find our stringid and grfid in the list, allocate a new id */
+		_grf_text.emplace_back();
+		it = std::end(_grf_text) - 1;
```
```suggestion
		it = _grf_text.emplace(std::end(_grf_text));
```
Maybe ?

PeterN replied:
Thanks, that's the construct I was trying to find but failing...

TrueBrain commented on src/newgrf_text.cpp:
```diff
@@ -645,6 +638,7 @@ const char *GetGRFStringFromGRFText(const GRFTextWrapper &text)
  */
 const char *GetGRFStringPtr(uint16_t stringid)
 {
+	assert(_grf_text.size() > stringid);
```
Doesn't the inverse make more sense? Stringid < size? As I think it should be >= otherwise, not?

It is late, pretty sure I am wrong about the =. The order still confuses my brain :D

PeterN committed:
Codechange: Allow using more than 65536 NewGRF string IDs.

NewGRF string allocation allowed up to 524288 strings, however stringid was passed as uint16_t which limits to 2^16.

TrueBrain commented on src/newgrf_text.cpp:
```diff
@@ -561,27 +560,21 @@ StringID AddGRFString(uint32_t grfid, uint16_t stringid, byte langid_to_add, boo
-	}
+	auto it = std::find_if(std::begin(_grf_text), std::end(_grf_text), [&grfid, &stringid](const GRFTextEntry &grf_text) { return grf_text.grfid == grfid && grf_text.stringid == stringid; });
+	if (it == std::end(_grf_text)) {
+		/* Too many strings allocated, return empty */
```
missing dot at end of sentence :)

TrueBrain commented on src/newgrf_text.cpp:
```diff
@@ -561,27 +560,21 @@ StringID AddGRFString(uint32_t grfid, uint16_t stringid, byte langid_to_add, boo
 
-	/* Too many strings allocated, return empty */
-	if (id == lengthof(_grf_text)) return STR_EMPTY;
+		/* We didn't find our stringid and grfid in the list, allocate a new id */
```
missing dot at end of sentence :)
